packages/client/src/sync/fetcher/fetcher.ts

```diff
@@ -225,17 +231,25 @@ export abstract class Fetcher<JobTask, JobResult, StorageItem> extends Readable
    * @param result job result
    */
   private success(job: Job<JobTask, JobResult, StorageItem>, result?: JobResult) {
-    if (job.state !== 'active') return
     let jobStr = this.jobStr(job, true)
+    if (job.state !== 'active') return
+
     let reenqueue = false
     let resultSet = ''
     if (result === undefined) {
       resultSet = 'undefined'
       reenqueue = true
     }
-    if (result !== undefined && (result as any).length === 0) {
-      resultSet = 'empty'
-      reenqueue = true
+    if (result !== undefined) {
+      if ('length' in (result as any)) {
+        if ((result as any).length === 0) {
+          resultSet = 'empty'
+          reenqueue = true
+        }
+      } else {
+        resultSet = 'unknown'
+        reenqueue = true
```

Comment:
This line here is what breaks `integration/lightsync.spec.ts`

Reply:
the test is fixed by deleting `reenqueue = true`

Reply:
the "unknown" result that the test deals with looks like:
```
{
    reqId: BigInt,
    bv: BigInt,
    headers: BlockHeader[]
}
```

Reply:
Which is a `BlockHeaderResult` from the `HeaderFetcher`.

So the test is also fixed like:

```
 } else if ('headers' in (result as any)) {
        // is a header fetcher job
      } else {
        resultSet = 'unknown'
        reenqueue = true
      }
```

Reply:
Actually --  `result` only ever shows up in the integration tests as `undefined` or as a `BlockHeaderResult`.

So we could delete the whole `if (result !== undefined)` bit all together and and all of our tests would pass.

The timeline on that file only goes back as far as 3 months ago when `lib` became `src`.  I assume that it was written that way for a good reason, and that we just aren't testing it?  Or it is totally unnecessary :shrug:

Reply:
Ok, I tracked the `if (result as any).length === 0` line to a commit by ryan from a couple years ago called **sync stability fixes** #1543.

Reply:
Hmm, the guide should here what behavior we would want "in reality" and not what fixes the tests. But you are correct, if the `reenqueue` line is deleted, the tests pass again.

For devp2p communication it is always good to try to make robust choices. So if the other side delivers something strange, just have a case for that. Here it would be that the structure `(result as any)` does not have a `length` property. The logic I applied here was that this would mean that the other side delivered something unepected (useless) which should therefore be thrown away and the task reenqueued (since it didn't resolve in a successful way). I would still think this should make sense. 🤔

Will have a look into the test case, maybe the assumptions are wong there. Also, the lightclient part of the client is really not maintained (and not even working). Maybe the conclusion here might also just be to throw these tests away.

Reply:
Have now added a lightsync specific hotfix in form of an extra conditional with a comment to eventually remove in the future.